### PR TITLE
Add retry and timeout to flaky DCDO test

### PR DESCRIPTION
**Pull request by carl-codeorg**
Adds a 1-second retry and 10-second timeout to a flaky DCDO test. Previously, [I had a PR up](https://github.com/code-dot-org/code-dot-org/pull/60021) to correct this same issue by making the call to DynamoDB strongly consistent, but the majority of reviewers did not like that solution.

This retry loop is a bit more basic of a solution, but carries no risk of affecting other tests or reducing parity with production configuration. Note that this test is only flaky in the test env (which uses DynamoDB to back DCDO), so we won't be able to observe that the test is truly fixed until it's been merged and no longer produces flaky results over time.

## Links

[JIRA](https://codedotorg.atlassian.net/browse/P20-1064)

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Comment by artem-vavilov** [comment]
We would see the error "DynamoDBAdapter does not support clear" if we use DynamoDB for unit testing:
https://github.com/code-dot-org/code-dot-org/blob/a6859c54dd045d99396375ab85c11388452141d0/lib/dynamic_config/adapters/dynamodb_adapter.rb#L78-L80

https://github.com/code-dot-org/code-dot-org/blob/a6859c54dd045d99396375ab85c11388452141d0/dashboard/test/test_helper.rb#L84

File: dashboard/test/lib/dcdo_test.rb [hunk 1]
```diff
@@ -19,7 +20,20 @@ class DCDOTest < ActiveSupport::TestCase
     }
     key = 'random'
     DCDO.set(key, to_store)
-    assert_equal DCDO.get(key, nil), to_store
```
Comment (artem-vavilov) [inline dashboard/test/lib/dcdo_test.rb:22]: The test "basic set and get" is quite similar, is it also flaky?
https://github.com/code-dot-org/code-dot-org/blob/a6859c54dd045d99396375ab85c11388452141d0/dashboard/test/lib/dcdo_test.rb#L6-L10

Comment (artem-vavilov) [reply dashboard/test/lib/dcdo_test.rb:22]: I have a suspicion that the issue might be in `DCDO.stubs(:get)`, which we use frequently in other test cases.

We could try unstubbing the method `DCDO.get` before each test:
```ruby
setup do
  DCDO.unstub(:get)
end

test 'basic set and get' do 
  assert_nil DCDO.get('key', nil) 
  DCDO.set('key', 'okay') 
  assert_equal DCDO.get('key', nil), 'okay' 
end 
```

or test a new instance of `DCDO` that is not stubbed:
```ruby
setup do
  @dcdo = DCDO.dup
end

test 'basic set and get' do
  assert_nil @dcdo.get('key', nil) 
  @dcdo.set('key', 'okay') 
  assert_equal @dcdo.get('key', nil), 'okay' 
end
```

Comment (artem-vavilov) [reply dashboard/test/lib/dcdo_test.rb:22]: We could also ensure that the `DCDO` will not be rewritten in any other place by using a unique key for each test:
```ruby
test 'basic set and get' do
  dcdo_key = SecureRandom.hex
  dcdo_value = 'okay'

  assert_changes -> {DCDO.get(dcdo_key, nil)}, from: nil, to: dcdo_value do
    DCDO.set(dcdo_key, dcdo_value)
  end
end
```

Comment (carl-codeorg) [reply dashboard/test/lib/dcdo_test.rb:22]: > The test "basic set and get" is quite similar, is it also flaky?
> 
> https://github.com/code-dot-org/code-dot-org/blob/a6859c54dd045d99396375ab85c11388452141d0/dashboard/test/lib/dcdo_test.rb#L6-L10

Good catch - I'm assuming it also has the potential to be flaky even though it hasn't been called out yet, so I added the retry there as well.

Comment (carl-codeorg) [reply dashboard/test/lib/dcdo_test.rb:22]: > I have a suspicion that the issue might be in `DCDO.stubs(:get)`, which we use frequently in other test cases.
> 

Good idea; I added the unstub

**Comment by carl-codeorg** [comment]
closing for now - don't want to take up reviewers time on this if there's a possible better solution

**Pull request closed without merging**